best-practices/client-side/framework/react/components/media-cropper/src/components/media-cropper/index.js

```diff
@@ -0,0 +1,184 @@
+import React, { Component } from 'react';
+
+import ImageRenderer from '../image-renderer';
+
+export default class MediaCropper extends Component {
+    constructor(props) {
+        super(props);
+        const media = new Image();
+        media.src = `https://via.placeholder.com/${props.width}x${props.height}.png?text=click%20to%20add`;
+
+        this.state = {
+            zoom: 1,
+            transX: 0,
+            transY: 0,
+            media,
+            isVideo: false,
+            dragging: false,
+            mouseStart: {
+                x: 0,
+                y: 0
+            },
+            moved: false
+        };
+
+        this.click = this.click.bind(this);
+        this.mouseDown = this.mouseDown.bind(this);
+        this.mouseMove = this.mouseMove.bind(this);
+        this.mouseUp = this.mouseUp.bind(this);
+        this.mouseWheel = this.mouseWheel.bind(this);
+        this.loadFile = this.loadFile.bind(this);
+    }
+
+    componentDidUpdate(props, state) {
+        if (this.state.dragging && !state.dragging) {
+            document.addEventListener('mousemove', this.mouseMove);
+            document.addEventListener('mouseup', this.mouseUp);
+            document.addEventListener('mouseleave', this.mouseUp);
+        } else if (!this.state.dragging && state.dragging) {
+            document.removeEventListener('mousemove', this.mouseMove);
+            document.removeEventListener('mouseup', this.mouseUp);
+            document.removeEventListener('mouseleave', this.mouseUp);
+        }
+    }
+
+    click(e) {
+        if (!this.state.moved) {
+            this.refs.fileUploader.click();
+        }
+        this.setState({
+            ...this.state,
+            moved: false
+        })
+    }
+
+    mouseDown(e) {
+        this.setState({
+            ...this.state,
+            dragging: true,
+            mouseStart: {
+                x: e.pageX,
+                y: e.pageY
+            }
+        });
+        e.preventDefault();
+        e.stopPropagation();
+    }
+
+    mouseMove(e) {
+        this.setState({
+            ...this.state,
+            transX: this.state.transX - (this.state.mouseStart.x - e.pageX),
+            transY: this.state.transY - (this.state.mouseStart.y - e.pageY),
+            mouseStart: {
+                x: e.pageX,
+                y: e.pageY
+            },
+            moved: true
+        });
+        e.preventDefault();
+        e.stopPropagation();
+    }
+
+    mouseUp(e) {
+        this.setState({
+            ...this.state,
+            dragging: false
+        });
+        e.preventDefault();
+        e.stopPropagation();
+    }
+
+    mouseWheel(e) {
+        const scale = Math.pow(1.0015, -e.deltaY);
```

Comment:
What is this indicating?

Reply:
this is used for zooming, first we need to know how much we are adjusting the zoom (that is the scale). then we can calculate how to adjust the translation based on the change in zoom. This allows us to "zoom around a point."